File: TrappingRainWater/TrappingRainWater.cpp
/*
 * time: O(n), space: O(1)
 */
#include <iostream>
#include <stack>
using namespace std;

class Solution{
public:
    int trap(int A[], int n){
        if(n == 0)
            return 0;

        int maxIndex = 0;
        for(int i = 0; i < n; i++){
            if(A[i] > A[maxIndex])
                maxIndex = i;
        }

        int sum = 0;
        
        int maxLeft = 0;
        for(int i = 0; i < maxIndex; i++){
            maxLeft = max(maxLeft, A[i]);
            sum += maxLeft - A[i];
        }

        int maxRight = 0;
        for(int i = n - 1; i > maxIndex; i--){
            maxRight = max(maxRight, A[i]);
            sum += maxRight - A[i];
        }

        return sum;
    }
};

int main(){
    int A[] = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};

    Solution solution;
    int res = solution.trap(A, sizeof(A) / sizeof(A[0]));
    cout << res << endl;

    return 0;
}

File: algorithms/TrappingRainWater/TrappingRainWater.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        if not height:
            return 0
        
        max_height = max(height)
        max_index = height.index(max_height)
        
        area_left, max_left = 0, height[0]
        for i in range(1, max_index):
            max_left = max(max_left, height[i])
            area_left += max_left - height[i]
            
        area_right, max_right = 0, height[-1]
        for i in range(len(height) - 2, max_index, -1):
            max_right = max(max_right, height[i])
            area_right += max_right - height[i]
        
        return area_left + area_right
